packages/xstate-solid/src/utils.ts

```diff
@@ -0,0 +1,73 @@
+import type { AnyState, AnyStateMachine, State, StateFrom } from 'xstate';
+import type { SetStoreFunction } from 'solid-js/store';
+import { produce, reconcile } from 'solid-js/store';
+import rfdc from 'rfdc';
+
+// List of keys to reconcile while merging state
+const reconcileKeys: Array<keyof AnyState> = ['context', 'value'];
+
+/**
+ * Reconcile the state of the machine with the current state of the store.
+ * Handles primitive values, arrays, and objects.
+ * Provides granular reactivity for the state of the machine in solid-js.
+ */
+export const updateState = <
+  NextState extends StateFrom<AnyStateMachine> | object,
+  UpdateState extends SetStoreFunction<NextState>
+>(
+  nextState: NextState,
+  setState: UpdateState,
+  keys?: Array<keyof NextState>
+): void => {
+  if (typeof nextState === 'object' && !!nextState) {
+    // If a list of keys to update is not provided, get keys from nextState
+    if (!keys) {
+      keys = Object.keys(nextState) as Array<keyof NextState>;
+    }
+    for (const key of keys) {
+      // Don't update functions
+      if (typeof nextState[key] === 'function') {
+        continue;
+      }
+      if (key in nextState && reconcileKeys.includes(key as keyof State<any>)) {
+        setReconcileState(key, nextState, setState);
+      } else {
+        setProduceState(key, nextState, setState);
```

Comment:
If I understand correctly, we are here only reconciling `context` and `value` but everything else is only reactively-tracked on the top-level of the store. Is this a common practice/tradeoff in Solid? Or would a common expectation be to have a "deep" reactivity?

Reply:
I mostly wonder about cases like this:
```js
const selected = useSelector(actorRef, newValue => ({ obj: newValue.obj })
selected.obj.deep.prop
```
What would be the expectation? Should this track `selected.obj` or should it try to track `selected.obj.deep.prop` for its granular changes?

Reply:
This is a good question, I would say most people would expect deep reactivity with Solid to a point. I think the most important is user defined data structures being deeply reactive. Further than that, users could be allowed to enable expanded reconciliation globally or on a per machine/actor basis. 

I've tried to err on the safe side with useSelector updates. In most situations, useSelector isn't necessary as you can just access a nested value in a machine or actor without worrying about re-renders. 

Overall, performance does not seem like a problem with reconciliation. I've done some very basic tests with deep nested and large updates with the scripting step only lasting an extra 20ms. 

Are there any other fields we should reconcile on updates? I can also enable reconciliation on non state objects.

Reply:
Solid's DOM doesn't diff so there would be expectations with any deep data. Reconcile from perf standpoint is similar to the fastest VDOMs, pulling numbers ultimately in the Inferno range.

Reply:
```typescript
// List of keys to reconcile while merging state
export const stateReconcileKeys: Array<keyof AnyState> = ['context'];

/**
 * Reconcile the state of the machine with the current state of the store.
 * Handles primitive values, arrays, and objects.
 * Provides granular reactivity for the state of the machine in solid-js.
 */
export function updateWithServiceState<NextState extends AnyState>(
  nextState: NextState,
  setState: SetStoreFunction<NextState>
): void {
  const keys = Object.keys(nextState) as Array<keyof NextState>;
  for (const key of keys) {
    if (stateReconcileKeys.includes(key as keyof AnyState)) {
      setState(
        key as any,
        reconcile<NextState[keyof NextState], unknown>(nextState[key])
      );
    } else {
      setState(produce((state) => (state[key as any] = nextState[key])));
    }
  }
}

export function updateState<NextState extends object>(
  nextState: NextState,
  setState: SetStoreFunction<NextState>
): void {
  if (nextState instanceof State) {
    updateWithServiceState(nextState, setState as SetStoreFunction<AnyState>);
  } else if (typeof nextState === 'object' && !!nextState) {
    setState(reconcile<NextState, unknown>(nextState));
  } else {
    setState(nextState);
  }
}
```

Here's an alternate implementation. Uses instanceof State to only loop through keys if the object passed is a State. If it's just an object, it reconciles from the top level. We remove looping for selectors and any actor snapshots. Providing deep reactivity by default

We could simplify further and remove updateWithServiceState. Handling State as well as anything from useActor or useSelector.

```typescript
export function updateState<NextState extends object>(
  nextState: NextState,
  setState: SetStoreFunction<NextState>
): void {
  if (typeof nextState === 'object' && !!nextState) {
    setState(reconcile<NextState, unknown>(nextState));
  } else {
    setState(nextState);
  }
}
```

Reply:
Based on @ryansolid's response, I feel like my last `updateState` snippet would be the best move going forward.